src/Sylius/Bundle/ApiBundle/Validator/Constraints/PromotionCouponEligibilityValidator.php

```diff
@@ -43,29 +47,44 @@ public function validate(mixed $value, Constraint $constraint): void
             return;
         }
 
+        /** @var PromotionCouponInterface|null $promotionCoupon */
         $promotionCoupon = $this->promotionCouponRepository->findOneBy(['code' => $value->couponCode]);
 
-        if (!$promotionCoupon instanceof PromotionCouponInterface) {
-            $this->context
-                ->buildViolation($constraint->message)
-                ->atPath('couponCode')
-                ->addViolation()
-            ;
+        if ($promotionCoupon === null) {
+            $this->addViolation($constraint->invalid, 'COUPON_INVALID');
 
             return;
         }
 
-        /** @var OrderInterface $cart */
+        $expirationDate = $promotionCoupon->getExpiresAt();
+
+        if ($expirationDate !== null && $expirationDate < new \DateTime()) {
```

Comment:
I think we're suppose to use `symfony/clock` for dates now.

```suggestion
        $now = $this->clock->now()
        if ($expiresAt !== null && $now > $expiresAt) {
```